File: 2023/16.2.py
import sys
import copy

try:
    sys.stdin = open(sys.path[0] + "/input.txt", "r")
    sys.stdout = open(sys.path[0] + "/output.txt", "w")
except FileNotFoundError:
    pass


def parse_input():
    grid = []
    for line in sys.stdin:
        grid.append(line.strip())

    return grid


def simulate_beam(grid, start_x, start_y, dx, dy):
    m, n = len(grid), len(grid[0])

    coordinates = {"L": (0, -1), "R": (0, 1), "T": (-1, 0), "B": (1, 0)}

    # This reads as:
    # If the beam is coming from (x, y) where (x, y) is the key
    # It can reflect in any of the direction_map[(x, y)] directions
    direction_map = {
        (0, -1): {
            "-": ("L"),
            "|": ("B", "T"),
            "/": ("B"),
            "\\": ("T"),
            ".": ("L"),
        },
        (0, 1): {
            "-": ("R"),
            "|": ("B", "T"),
            "/": ("T"),
            "\\": ("B"),
            ".": ("R"),
        },
        (-1, 0): {
            "-": ("L", "R"),
            "|": ("T"),
            "/": ("R"),
            "\\": ("L"),
            ".": ("T"),
        },
        (1, 0): {
            "-": ("L", "R"),
            "|": ("B"),
            "/": ("L"),
            "\\": ("R"),
            ".": ("B"),
        },
    }

    track = [["." for _ in range(n)] for __ in range(m)]

    # x, y, dx, dy
    # Stack because it is memory efficient.
    stack = [(start_x, start_y, dx, dy)]

    visited = set()

    while stack:
        key = stack.pop()
        x, y, dx, dy = key

        if key in visited:
            continue
        visited.add(key)
        track[x][y] = "#"

        curr = grid[x][y]
        for ch in direction_map[(dx, dy)][curr]:
            ndx, ndy = coordinates[ch]
            if 0 <= x + ndx < m and 0 <= y + ndy < n:
                stack.append((x + ndx, y + ndy, ndx, ndy))

    return track


def find_max_energized_tiles(grid):
    max_energized_tiles = 0

    # Check for leftmost and rightmost column
    for i in range(len(grid)):
        max_energized_tiles = max(
            max_energized_tiles,
            count_energized_tiles(simulate_beam(grid, i, 0, 0, 1)),
            count_energized_tiles(simulate_beam(grid, i, len(grid[0]) - 1, 0, -1)),
        )

    # Check for top and bottom rows
    for j in range(len(grid[0])):
        max_energized_tiles = max(
            max_energized_tiles,
            count_energized_tiles(simulate_beam(grid, 0, j, 1, 0)),
            count_energized_tiles(simulate_beam(grid, len(grid) - 1, j, -1, 0)),
        )

    return max_energized_tiles


def count_energized_tiles(grid):
    cnt = 0
    for row in grid:
        cnt += row.count("#")

    return cnt


grid = parse_input()

print("Ans:", find_max_energized_tiles(grid))

File: 2023/16.2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <stack>
#include <map>
#include <tuple>
#include <unordered_set>
#include <algorithm>

using namespace std;

ofstream outputFile;
const string inputFileName = "input.txt";
const string outputFileName = "output.txt";

struct TupleHashFunction
{
    size_t operator()(const tuple<int, int, int, int> &x) const
    {
        return get<0>(x) ^ get<1>(x) ^ get<2>(x) ^ get<3>(x);
    }
};

void setOutputStream()
{
    outputFile.open(outputFileName);
    if (!outputFile.is_open())
        exit(EXIT_FAILURE);

    cout.rdbuf(outputFile.rdbuf());
}

void displayGrid(vector<vector<char>> &grid)
{
    for (auto &row : grid)
    {
        for (auto &ch : row)
        {
            cout << ch;
        }
        cout << endl;
    }
}

vector<vector<char>> parseInput()
{
    ifstream inputFile(inputFileName);
    if (!inputFile.is_open())
    {
        exit(EXIT_FAILURE);
    }

    string line;
    vector<vector<char>> grid;
    while (getline(inputFile, line))
    {
        vector<char> singleLine;
        for (auto &ch : line)
        {
            singleLine.push_back(ch);
        }

        grid.push_back(singleLine);
    }

    return grid;
}

vector<vector<char>> simulateBeamPath(vector<vector<char>> &grid, int start_x, int start_y, int dx, int dy)
{
    const size_t M = grid.size(), N = grid[0].size();

    map<char, pair<int, int>> coordinates = {
        {'L', {0, -1}},
        {'R', {0, 1}},
        {'T', {-1, 0}},
        {'B', {1, 0}}};

    map<pair<int, int>, map<char, vector<char>>> directionMap = {
        {{0, -1}, {{'-', {'L'}}, {'|', {'B', 'T'}}, {'/', {'B'}}, {'\\', {'T'}}, {'.', {'L'}}}},
        {{0, 1}, {{'-', {'R'}}, {'|', {'B', 'T'}}, {'/', {'T'}}, {'\\', {'B'}}, {'.', {'R'}}}},
        {{-1, 0}, {{'-', {'L', 'R'}}, {'|', {'T'}}, {'/', {'R'}}, {'\\', {'L'}}, {'.', {'T'}}}},
        {{1, 0}, {{'-', {'L', 'R'}}, {'|', {'B'}}, {'/', {'L'}}, {'\\', {'R'}}, {'.', {'B'}}}}};

    vector<vector<char>> track(M, vector<char>(N, '.'));

    stack<tuple<int, int, int, int>> dfsStack;
    // start_x, start_y, dx, dy
    dfsStack.push(make_tuple(start_x, start_y, dx, dy));
    unordered_set<tuple<int, int, int, int>, TupleHashFunction> visited;

    while (!dfsStack.empty())
    {
        int x, y, dx, dy;
        tuple<int, int, int, int> front = dfsStack.top();
        tie(x, y, dx, dy) = front;
        dfsStack.pop();

        if (visited.find(front) != visited.end())
        {
            continue;
        }
        visited.insert(front);
        track[x][y] = '#';

        char curr = grid[x][y];
        for (auto &ch : directionMap[make_pair(dx, dy)][curr])
        {
            int ndx, ndy;
            tie(ndx, ndy) = coordinates[ch];
            if (0 <= x + ndx && x + ndx < M && 0 <= y + ndy && y + ndy < N)
            {
                dfsStack.push(make_tuple(x + ndx, y + ndy, ndx, ndy));
            }
        }
    }

    return track;
}

int countEnergizedTiles(vector<vector<char>> &track)
{
    int cnt = 0;
    for (auto &row : track)
    {
        cnt += count(row.begin(), row.end(), '#');
    }

    return cnt;
}

int findMaxPossibleEnergizedTiles(vector<vector<char>> &grid)
{
    int maxTiles = 0;
    vector<vector<char>> tempTrack;

    // Check left and right
    for (size_t i = 0; i < grid.size(); i++)
    {
        tempTrack = simulateBeamPath(grid, i, 0, 0, 1);
        maxTiles = max(maxTiles, countEnergizedTiles(tempTrack));

        tempTrack = simulateBeamPath(grid, i, grid.size() - 1, 0, -1);
        maxTiles = max(maxTiles, countEnergizedTiles(tempTrack));
    }

    // Check top and bottom
    for (size_t j = 0; j < grid[0].size(); j++)
    {
        tempTrack = simulateBeamPath(grid, 0, j, 1, 0);
        maxTiles = max(maxTiles, countEnergizedTiles(tempTrack));

        tempTrack = simulateBeamPath(grid, grid[0].size() - 1, j, -1, 0);
        maxTiles = max(maxTiles, countEnergizedTiles(tempTrack));
    }

    return maxTiles;
}

int main()
{
    setOutputStream();

    vector<vector<char>> grid = parseInput();
    // displayGrid(grid);

    cout << "Ans: " << findMaxPossibleEnergizedTiles(grid) << endl;

    return 0;
}